src/main/java/com/yscope/clp/compressorfrontend/MessageDecoder.java

```diff
@@ -58,34 +60,89 @@ public String decodeMessage(
       @NotNull String logtype,
       String[] dictionaryVars,
       long[] encodedVars
+  ) throws IOException {
+    byte[] messageBytes = decodeMessageBytes(logtype, dictionaryVars, encodedVars);
+    return new String(messageBytes, StandardCharsets.UTF_8);
+  }
+
+  /**
+   * Decodes the message with the given logtype and variables
+   * @param logtypeBytes
+   * @param flattenedDictionaryVarsBytes
+   * @param flattenedDictionaryVarEndOffsets
+   * @param encodedVars
+   * @return The decoded message
+   * @throws IOException if decoding fails
+   */
+  public String decodeMessage(
+      @NotNull byte[] logtypeBytes,
+      byte[] flattenedDictionaryVarsBytes,
+      int[] flattenedDictionaryVarEndOffsets,
+      long[] encodedVars
+  ) throws IOException {
+    byte[] messageBytes = decodeMessageBytes(logtypeBytes, flattenedDictionaryVarsBytes,
+        flattenedDictionaryVarEndOffsets, encodedVars);
+    return new String(messageBytes, StandardCharsets.UTF_8);
+  }
+
+  /**
+   * Decodes the message bytes with the given logtype and variables
+   * @param logtype
+   * @param dictionaryVars
+   * @param encodedVars
+   * @return The decoded message
+   * @throws IOException if decoding fails
+   */
+  private byte[] decodeMessageBytes(
+      @NotNull String logtype,
+      String[] dictionaryVars,
+      long[] encodedVars
   ) throws IOException {
     Objects.requireNonNull(logtype);
+    byte[] logtypeBytes = logtype.getBytes(StandardCharsets.ISO_8859_1);
 
     int[] dictionaryVarEndOffsets = null;
     byte[] allDictionaryVars = null;
     if (null != dictionaryVars) {
       // Flatten dictionaryVars (["var1", "var2", ...] -> "var1var2...")
       dictionaryVarEndOffsets = new int[dictionaryVars.length];
-      ByteArrayOutputStream dictionaryVarsOutputStream = new ByteArrayOutputStream();
+      flattenedDictionaryVarsBytesOutputStream = new ByteArrayOutputStream();
```

Comment:
Have you measured whether promoting flattenedDictionaryVarsBytesOutputStream has any effect on heap usage? Since the code is still creating a new ByteArrayOutputStream every time, I don't think any experiments would've shown a difference?